src/Functions/array/arrayShiftRotate.cpp

```diff
@@ -0,0 +1,419 @@
+    }
+
+    void insertManyToColumn(IColumn & dst, size_t row_num, size_t length) const
+    {
+        if (column)
+            dst.insertManyFrom(*column, row_num, length);
+        else
+            dst.insertMany(const_value, length);
+    }
+
+    Field const_value;
+    ColumnPtr column;
+};
+
+template <typename Impl, typename Name>
+class FunctionArrayShiftRotate : public IFunction
+{
+public:
+    static constexpr auto name = Name::name;
+    static constexpr ShiftRotateStrategy strategy = Impl::strategy;
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionArrayShiftRotate>(); }
+
+    String getName() const override { return name; }
+
+    bool isVariadic() const override { return strategy == ShiftRotateStrategy::Shift; }
+    size_t getNumberOfArguments() const override { return strategy == ShiftRotateStrategy::Rotate ? 2 : 0; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if constexpr (strategy == ShiftRotateStrategy::Shift)
+        {
+            if (arguments.size() < 2)
+                throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires at least two arguments.", getName());
+
+            if (arguments.size() > 3)
+                throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires at most three arguments.", getName());
+        }
+
+        const DataTypePtr & first_arg = arguments[0];
+        if (!isArray(first_arg))
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Illegal type {} of argument of function {}, expected Array",
+                arguments[0]->getName(),
+                getName());
+
+        if (!isNativeInteger(arguments[1]))
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Illegal type {} of argument of function {}, expected Native Integer",
+                arguments[1]->getName(),
+                getName());
+
+        const DataTypePtr & elem_type = static_cast<const DataTypeArray &>(*first_arg).getNestedType();
+        if (arguments.size() == 3)
+        {
+            auto ret = tryGetLeastSupertype(DataTypes{elem_type, arguments[2]});
+            // Note that this will fail if the default value does not fit into the array element type (e.g. UInt64 and Array(UInt8)).
+            // In this case array should be converted to Array(UInt64) explicitly.
+            if (!ret || !ret->equals(*elem_type))
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Illegal type {} of argument of function {}, expected {}",
+                    arguments[2]->getName(),
+                    getName(),
+                    elem_type->getName());
```

Comment:
If it's possible, consider adding test cases to check this cases also. Especially last one with different types.

Reply:
Sure! Will do